File: client/lib/geocoding_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'url_config_service.dart';

class GeocodingService {
  static const int _defaultCount = 10;
  static const int _maxCount = 50;

  // City/Location models
  static GeocodingResult? _selectedCity;
  static TimezoneInfo? _selectedTimezone;
  
  static GeocodingResult? get selectedCity => _selectedCity;
  static TimezoneInfo? get selectedTimezone => _selectedTimezone;
  
  static void setSelectedCity(GeocodingResult? city) {
    _selectedCity = city;
  }
  
  static void setSelectedTimezone(TimezoneInfo? timezone) {
    _selectedTimezone = timezone;
  }

  // Search for cities using the server's geocoding API
  static Future<List<GeocodingResult>> searchCities(String query, {int count = _defaultCount}) async {
    if (query.length < 2) {
      return [];
    }

    final serverUrl = await UrlConfigService.getServerUrl();
    if (serverUrl == null || serverUrl.isEmpty) {
      throw Exception('No server URL configured');
    }

    try {
      final uri = Uri.parse('$serverUrl/api/geocoding/search')
          .replace(queryParameters: {
        'name': query,
        'count': count.clamp(1, _maxCount).toString(),
      });

      final response = await http.get(uri).timeout(
        const Duration(seconds: 10),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final results = data['results'] as List<dynamic>? ?? [];
        
        return results.map((json) => GeocodingResult.fromJson(json)).toList();
      } else if (response.statusCode == 400) {
        // Handle bad request (e.g., query too short)
        return [];
      } else {
        throw Exception('Geocoding API returned status ${response.statusCode}');
      }
    } catch (e) {
      print('Error searching cities: $e');
      throw Exception('Failed to search cities: $e');
    }
  }

  // Search for timezones using the server's timezone API
  static Future<List<TimezoneInfo>> searchTimezones(String query, {int limit = 50}) async {
    final serverUrl = await UrlConfigService.getServerUrl();
    if (serverUrl == null || serverUrl.isEmpty) {
      throw Exception('No server URL configured');
    }

    try {
      final uri = Uri.parse('$serverUrl/api/timezones/search')
          .replace(queryParameters: {
        'q': query,
        'limit': limit.toString(),
      });

      final response = await http.get(uri).timeout(
        const Duration(seconds: 10),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final results = data['results'] as List<dynamic>? ?? [];
        
        return results.map((json) => TimezoneInfo.fromJson(json)).toList();
      } else {
        throw Exception('Timezone API returned status ${response.statusCode}');
      }
    } catch (e) {
      print('Error searching timezones: $e');
      throw Exception('Failed to search timezones: $e');
    }
  }

  // Get all available timezones
  static Future<List<TimezoneInfo>> getAllTimezones() async {
    final serverUrl = await UrlConfigService.getServerUrl();
    if (serverUrl == null || serverUrl.isEmpty) {
      throw Exception('No server URL configured');
    }

    try {
      final response = await http.get(
        Uri.parse('$serverUrl/api/timezones'),
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final results = data['results'] as List<dynamic>? ?? [];
        
        return results.map((json) => TimezoneInfo.fromJson(json)).toList();
      } else {
        throw Exception('Timezone API returned status ${response.statusCode}');
      }
    } catch (e) {
      print('Error getting all timezones: $e');
      throw Exception('Failed to get timezones: $e');
    }
  }
}

class GeocodingResult {
  final int id;
  final String name;
  final double latitude;
  final double longitude;
  final double elevation;
  final String featureCode;
  final String countryCode;
  final int countryId;
  final String country;
  final String timezone;
  final int population;
  final String? admin1;
  final String? admin2;
  final String? admin3;
  final String? admin4;

  GeocodingResult({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.elevation,
    required this.featureCode,
    required this.countryCode,
    required this.countryId,
    required this.country,
    required this.timezone,
    required this.population,
    this.admin1,
    this.admin2,
    this.admin3,
    this.admin4,
  });

  factory GeocodingResult.fromJson(Map<String, dynamic> json) {
    return GeocodingResult(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      elevation: (json['elevation'] ?? 0.0).toDouble(),
      featureCode: json['feature_code'] ?? '',
      countryCode: json['country_code'] ?? '',
      countryId: json['country_id'] ?? 0,
      country: json['country'] ?? '',
      timezone: json['timezone'] ?? 'UTC',
      population: json['population'] ?? 0,
      admin1: json['admin1'],
      admin2: json['admin2'],
      admin3: json['admin3'],
      admin4: json['admin4'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'latitude': latitude,
      'longitude': longitude,
      'elevation': elevation,
      'feature_code': featureCode,
      'country_code': countryCode,
      'country_id': countryId,
      'country': country,
      'timezone': timezone,
      'population': population,
      'admin1': admin1,
      'admin2': admin2,
      'admin3': admin3,
      'admin4': admin4,
    };
  }

  // Display name for dropdown
  String get displayName {
    final parts = <String>[name];
    if (admin1 != null && admin1!.isNotEmpty) {
      parts.add(admin1!);
    }
    parts.add(country);
    return parts.join(', ');
  }

  // Short display name for selected items
  String get shortDisplayName {
    return '$name, $country';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is GeocodingResult && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => displayName;
}

class TimezoneInfo {
  final String id;
  final String displayName;
  final String region;
  final String city;

  TimezoneInfo({
    required this.id,
    required this.displayName,
    required this.region,
    required this.city,
  });

  factory TimezoneInfo.fromJson(Map<String, dynamic> json) {
    return TimezoneInfo(
      id: json['id'] ?? '',
      displayName: json['display_name'] ?? '',
      region: json['region'] ?? '',
      city: json['city'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'display_name': displayName,
      'region': region,
      'city': city,
    };
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is TimezoneInfo && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => displayName;
}

File: server/services/geocoding_service.go
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"

	"github.com/azukaar/sumika/server/utils"
)

type GeocodingService struct {
	cache         map[string]*GeocodingResponse
	cacheDuration time.Duration
}

type GeocodingResponse struct {
	Results []GeocodingResult `json:"results"`
	Cached  time.Time         `json:"-"`
}

type GeocodingResult struct {
	ID            int     `json:"id"`
	Name          string  `json:"name"`
	Latitude      float64 `json:"latitude"`
	Longitude     float64 `json:"longitude"`
	Elevation     float64 `json:"elevation"`
	FeatureCode   string  `json:"feature_code"`
	CountryCode   string  `json:"country_code"`
	CountryID     int     `json:"country_id"`
	Country       string  `json:"country"`
	Timezone      string  `json:"timezone"`
	Population    int     `json:"population"`
	Admin1ID      int     `json:"admin1_id,omitempty"`
	Admin1        string  `json:"admin1,omitempty"`
	Admin2ID      int     `json:"admin2_id,omitempty"`
	Admin2        string  `json:"admin2,omitempty"`
	Admin3ID      int     `json:"admin3_id,omitempty"`
	Admin3        string  `json:"admin3,omitempty"`
	Admin4ID      int     `json:"admin4_id,omitempty"`
	Admin4        string  `json:"admin4,omitempty"`
}

func NewGeocodingService() *GeocodingService {
	return &GeocodingService{
		cache:         make(map[string]*GeocodingResponse),
		cacheDuration: 24 * time.Hour, // Cache geocoding results for 24 hours
	}
}

// SearchCities searches for cities using Open-Meteo's Geocoding API
func (s *GeocodingService) SearchCities(query string, count int) (*GeocodingResponse, error) {
	context := utils.NewLogContext("geocoding").WithOperation("search_cities")
	
	// Validate query
	if len(query) < 2 {
		return nil, fmt.Errorf("query must be at least 2 characters long")
	}
	
	// Check cache first
	cacheKey := fmt.Sprintf("%s-%d", query, count)
	if cached, exists := s.cache[cacheKey]; exists {
		if time.Since(cached.Cached) < s.cacheDuration {
			utils.InfoWithContext("Returning cached geocoding results", context)
			return cached, nil
		}
		// Remove expired cache entry
		delete(s.cache, cacheKey)
	}

	// Build API URL
	baseURL := "https://geocoding-api.open-meteo.com/v1/search"
	params := url.Values{}
	params.Add("name", query)
	if count > 0 {
		params.Add("count", fmt.Sprintf("%d", count))
	}
	apiURL := fmt.Sprintf("%s?%s", baseURL, params.Encode())

	// Make API request
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(apiURL)
	if err != nil {
		utils.ErrorWithContext("Failed to fetch geocoding data", context, err)
		return nil, fmt.Errorf("failed to fetch geocoding data: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("geocoding API returned status %d", resp.StatusCode)
	}

	// Parse response
	var geocodingResp GeocodingResponse
	if err := json.NewDecoder(resp.Body).Decode(&geocodingResp); err != nil {
		utils.ErrorWithContext("Failed to decode geocoding response", context, err)
		return nil, fmt.Errorf("failed to decode geocoding response: %w", err)
	}

	// Cache the results
	geocodingResp.Cached = time.Now()
	s.cache[cacheKey] = &geocodingResp

	utils.InfoWithContext(fmt.Sprintf("Found %d geocoding results for query '%s'", len(geocodingResp.Results), query), context)
	return &geocodingResp, nil
}

// GetCityByID gets a specific city by its ID (useful for getting full details)
func (s *GeocodingService) GetCityByID(id int) (*GeocodingResult, error) {
	// For now, we don't have a direct ID lookup API, so this would require
	// implementing a local database or using a different approach
	return nil, fmt.Errorf("city lookup by ID not implemented")
}

// CleanCache removes expired cache entries
func (s *GeocodingService) CleanCache() {
	context := utils.NewLogContext("geocoding").WithOperation("clean_cache")
	
	now := time.Now()
	cleaned := 0
	
	for key, cached := range s.cache {
		if now.Sub(cached.Cached) >= s.cacheDuration {
			delete(s.cache, key)
			cleaned++
		}
	}
	
	if cleaned > 0 {
		utils.InfoWithContext(fmt.Sprintf("Cleaned %d expired geocoding cache entries", cleaned), context)
	}
}